test/NuGet.Clients.Tests/NuGet.VisualStudio.Implementation.Test/Extensibility/NuGetProjectServiceTests.cs

```diff
@@ -73,5 +81,114 @@ public async Task GetInstalledPackagesAsync_LegacyProjectInvalidDataException_Re
             Assert.NotNull(actual);
             Assert.Equal(InstalledPackageResultStatus.ProjectInvalid, actual.Status);
         }
+
+        [Fact]
+        public async Task GetInstalledPackagesAsync_PackageReferenceProject_ReturnsTransitive()
+        {
+            // Arrange
+            var projectGuid = Guid.NewGuid();
+
+            var settings = new Mock<ISettings>();
+            var telemetryProvider = new Mock<INuGetTelemetryProvider>(MockBehavior.Strict);
+
+            var installedPackages = new List<PackageReference>()
+            {
+                new PackageReference(new PackageIdentity("a", new Versioning.NuGetVersion(1, 0, 0)), FrameworkConstants.CommonFrameworks.Net50)
+            };
+            var transitivePackages = new List<PackageReference>()
+            {
+                new PackageReference(new PackageIdentity("b", new Versioning.NuGetVersion(1, 2, 3)), FrameworkConstants.CommonFrameworks.Net50)
+            };
+            var projectPackages = new ProjectPackages(installedPackages, transitivePackages);
+            var project = new TestPackageReferenceProject("ProjectA", @"src\ProjectA\Project.csproj", @"c:\path\to\src\ProjectA\ProjectA.csproj",
```

Comment:
I think you can save some testing code by reusing existing methods:

CpsPackageReferenceProject: https://github.com/NuGet/NuGet.Client/blob/b8fe2a0f84a2a1a7f4efe2dcaaa7876bef8f4627/test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/CpsPackageReferenceProjectTests.cs#L3293-L3293

LegacyPackageReferenceProject: 
https://github.com/NuGet/NuGet.Client/blob/b8fe2a0f84a2a1a7f4efe2dcaaa7876bef8f4627/test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs#L1337-L1337

I suggest move those methods into a static test class, like so:
https://github.com/NuGet/NuGet.Client/blob/adb5d8806476deb42be9b216bb62645c4d17bfdc/test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectFactories.cs

Reply:
I looked into TestCpsPackageReferenceProject, but it doesn't have extensibility to add transitive packages, only directly installed packages, which can then be retrieved by `GetPackageSpecs()`. What's worse is that CpsPackageReferenceProject, the production class the test class extends, it's implementation of `GetInstalledAndTransitivePackagesAsync()` will read the assets file. Therefore, in order to use any test class that extends `CpsPackageReferenecProject`, it's necessary to write a real assets file to disk, so that `GetInstalledAndTransitivePackagesAsync` can then read it.

On the other hand by extending `PackageReferenceProject`, I can just mock out a super simple `GetInstalledAndTransitivePackagesAsync`. Check both test class implementations. The one I added in this class is simpler. We really should try to make our production code more testable to reduce the complexity for us to write these types of tests.